addons/account/models/account_report.py

```diff
@@ -338,6 +338,27 @@ def _inverse_aggregation_formula(self):
     def _inverse_account_codes_formula(self):
         self._create_report_expression(engine='account_codes')
 
+    @api.ondelete(at_uninstall=False)
+    def _unlink_archive_used_tags(self):
+        """
+        Manages unlink or archive of tax_tags when a report.line is deleted.
+        If it's still in use either on amls or on other expressions
+        """
+        for expression in self.expression_ids:
+            if expression.engine == 'tax_tags':
+                country = expression.report_line_id.report_id.country_id
+                tags = self.env['account.account.tag']._get_tax_tags(expression.formula, country.id)
+                for tag in tags:
+                    expression_using_tag = self.env['account.report.expression'].sudo().search([('engine', '=', 'tax_tags'), ('formula', '=', tag.name)], limit=1)
+                    if not expression_using_tag:
+                        rep_lines_with_archived_tags = self.env['account.tax.repartition.line'].sudo().search([('tag_ids', 'in', tag.id)])
```

Comment:
You shouldn't do this in a loop. Let's batch things. For the search() above, you have no choice, but for this one, you could build a set of all the tags to delete in the loop, and then, after the loop, search all the repartition lines using them in one go, and remove them with a single write(). 

For the amls check, you have no choice, you need to keep it in the loop. The unlinking and archiving of the tag could be batched as well (if it does not make the code completely unreadable ; it's not worth making it weird-looking)